myokit/formats/diffsl/_exporter.py

```diff
@@ -0,0 +1,523 @@
+        export_lines.append('}')
+        export_lines.append('')
+
+        # Add `G_i`
+        export_lines.append('G_i {')
+        for v in model.states():
+            rhs = e.ex(v.rhs().rhs())  # Use rhs of intermediary dot_x
+            export_lines.append(f'{tab}{rhs},')
+        export_lines.append('}')
+        export_lines.append('')
+
+        # Output state variables + extra output variables
+        export_lines.append('/* Output */')
+        export_lines.append('out_i {')
+        vars = set(model.states())
+        if extra_out_vars:
+            vars = vars.union(extra_out_vars)
+        for v in sorted(vars, key=lambda x: x.qname()):
+            lhs = var_name(v)
+            export_lines.append(f'{tab}{lhs},')
+        export_lines.append('}')
+        export_lines.append('')
+
+        return '\n'.join(export_lines)
+
+    def _convert_current_unit(self, var, helpers=None):
+        """
+        Convert a current to A/F if its present unit isn't recommended.
+        """
+        recommended_units = [
+            myokit.parse_unit('pA/pF'),
+            myokit.parse_unit('uA/cm^2'),
+        ]
+
+        if var.unit() not in recommended_units:
+            self._convert_unit(var, 'A/F', helpers=helpers)
+
+    def _convert_potential_unit(self, var, helpers=None):
+        """
+        Convert a potential to mV if its present unit isn't recommended.
+        """
+        recommended_units = [myokit.units.mV]
+
+        if var.unit() not in recommended_units:
+            self._convert_unit(var, 'mV', helpers=helpers)
+
+    def _convert_unit(self, var, unit, helpers=None):
+        """
+        Convert a variable to the given unit if possible. Throws a warning if
+        the conversion is not possible.
+        """
+        if var.unit() is None:
+            return
+
+        try:
+            var.convert_unit(unit, helpers=helpers)
+        except myokit.IncompatibleUnitError:
+            warnings.warn(
+                'Unable to convert ' + var.qname() + ' to recommended'
+                ' units of ' + str(unit) + '.'
+            )
+
+    def _create_diffsl_variable_names(self, model):
+        """
+        Create diffsl-compatible names for all variables in the model.
+
+        The following strategy is followed:
+         - Fully qualified names are used for all variables.
+         - Variables are checked for special names, and changed if necessary.
+         - Unsupported characters like '.' and '_' are replaced.
+         - Any conflicts are resolved in a final step by appending a number.
+        """
+
+        # Detect names with special meanings
+        def special_start(name):
+            prefixes = ['dudt', 'F', 'G', 'in', 'M', 'out', 'u']
+            for prefix in prefixes:
+                if name.startswith(prefix + '_'):
+                    return True
+            return False
+
+        # Replacement characters for underscores and periods
+        uscore = 'Z'
+        period = 'Z'
```

Comment:
Is there some nicer strategy that gives more readable results? E.g. camelcasing?

It might also be nice to go from the `var.uname()`, which omits the component if not necessary for uniqueness. However, that's a personal preference so entirely up to you!

Reply:
Just seen `uname` is unique. Much easier, thanks.

Reply:
Resolved by 1a212493. Naming now starts from `qname()` and staggers case between unsupported characters e.g. `voltage_clamp.R_seal_MOhm` -> `voltageClampRSealMOhm`.